src/rust/ide/enso-protocol/src/file_manager.rs

```diff
@@ -94,96 +114,169 @@ pub enum FilesystemEventKind {
 // ==================
 
 /// Attributes of the file in the filesystem.
-#[derive(Clone,Copy,Debug,PartialEq)]
+#[derive(Clone,Debug,PartialEq,Eq,Hash)]
 #[derive(Serialize, Deserialize)]
 #[serde(rename_all = "camelCase")]
 pub struct Attributes{
     /// When the file was created.
-    pub creation_time      : UTCDateTime,
+    pub creation_time : UTCDateTime,
     /// When the file was last accessed.
-    pub last_access_time   : UTCDateTime,
+    pub last_access_time : UTCDateTime,
     /// When the file was last modified.
     pub last_modified_time : UTCDateTime,
     /// What kind of file is this.
-    pub file_kind          : FileKind,
+    pub kind : FileSystemObject,
     /// Size of the file in bytes.
     /// (size of files not being `RegularFile`s is unspecified).
-    pub byte_size          : u64
+    pub byte_size : u64
 }
 
-/// What kind of file (regular, directory, symlink) is this.
-#[derive(Clone,Copy,Debug,PartialEq)]
-#[derive(Serialize, Deserialize)]
-pub enum FileKind {
-    /// File being a directory.
-    Directory,
-    /// File being a symbolic link.
-    SymbolicLink,
-    /// File being a regular file with opaque content.
-    RegularFile,
-    /// File being none of the above, e.g. a physical device or a pipe.
-    Other
+/// A representation of what kind of type a filesystem object can be.
+#[derive(Hash,Debug,Clone,PartialEq,Eq,Serialize,Deserialize)]
+#[serde(tag = "type")]
+pub enum FileSystemObject {
+    /// Represents a directory.
+    Directory(Object),
+    /// Represents a directory which contents have been truncated
+    DirectoryTruncated(Object),
+    /// Represents a file.
+    File(Object),
+    /// Represents unrecognized object. Example is a broken symbolic link.
+    Other(Object),
+    /// Represents a symbolic link that creates a loop.
+    SymlinkLoop(SymlinkLoop)
+}
+
+/// Represents an object.
+#[derive(Hash,Debug,Clone,PartialEq,Eq,Serialize,Deserialize)]
+pub struct Object {
```

Comment:
What is `file_manager::Object`? The documentation does not tell what it is. I believe it should be moved to other module